File: node/pgn_benchmark.js
#!/usr/bin/env node
import { Chess } from 'chess.js';
import { parseArgs } from 'util';

function parsePgnMoves(pgn) {
  const match = pgn.match(/\n\n([\s\S]+)$/);
  if (!match) return [];
  let str = match[1].replace(/\{[^}]*\}/g, '').replace(/\([^)]*\)/g, '')
    .replace(/\d+\.\.\./g, '').replace(/\d+\./g, '').replace(/(1-0|0-1|1\/2-1\/2|\*)$/g, '');
  return str.trim().split(/\s+/).filter(m => m && !m.match(/^[\d.]+$/));
}

async function fetchArchives(username) {
  const resp = await fetch(`https://api.chess.com/pub/player/${username}/games/archives`,
    { headers: { 'User-Agent': 'ChessBenchmark/1.0' } });
  return (await resp.json()).archives || [];
}

async function fetchGames(url) {
  const resp = await fetch(url, { headers: { 'User-Agent': 'ChessBenchmark/1.0' } });
  return (await resp.json()).games || [];
}

function parseGame(pgn) {
  if (!pgn) return { moves: 0, positions: 0 };
  try {
    const chess = new Chess();
    const moves = parsePgnMoves(pgn);
    let moveCount = 0, posCount = 1;
    for (const m of moves) {
      try { chess.move(m); moveCount++; posCount++; chess.fen(); } catch { break; }
    }
    return { moves: moveCount, positions: posCount };
  } catch { return { moves: 0, positions: 0 }; }
}

async function main() {
  const { positionals } = parseArgs({ allowPositionals: true, options: {} });
  const username = positionals[0] || 'hikaru';
  const maxGames = parseInt(positionals[1]) || 1000;

  console.log('Node.js PGN Parsing Benchmark');
  console.log('='.repeat(50));
  console.log(`Library: chess.js`);
  console.log(`Username: ${username}`);
  console.log(`Max games: ${maxGames}\n`);

  console.log('Fetching games...');
  const fetchStart = performance.now();
  const archives = (await fetchArchives(username)).reverse();
  
  const allPgns = [];
  for (const url of archives) {
    if (allPgns.length >= maxGames) break;
    const games = await fetchGames(url);
    games.forEach(g => g.pgn && allPgns.push(g.pgn));
    console.log(`  Fetched ${games.length} games from ${url.split('/').slice(-2).join('/')}`);
  }

  const pgns = allPgns.slice(0, maxGames);
  console.log(`Fetched ${pgns.length} games in ${((performance.now() - fetchStart)/1000).toFixed(2)}s\n`);

  console.log('Parsing PGNs...');
  const parseStart = performance.now();
  let totalMoves = 0, totalPos = 0, parsed = 0;

  for (let i = 0; i < pgns.length; i += 100) {
    const batch = pgns.slice(i, i + 100);
    for (const pgn of batch) {
      const { moves, positions } = parseGame(pgn);
      if (moves > 0) { totalMoves += moves; totalPos += positions; parsed++; }
    }
    const elapsed = (performance.now() - parseStart) / 1000;
    console.log(`  Parsed ${Math.min(i + 100, pgns.length)}/${pgns.length} games (${((i+100)/elapsed).toFixed(2)} games/sec)`);
  }

  const parseTime = (performance.now() - parseStart) / 1000;
  console.log('\nResults');
  console.log('='.repeat(50));
  console.log(`Games parsed: ${parsed}`);
  console.log(`Total moves: ${totalMoves}`);
  console.log('\nPerformance');
  console.log('='.repeat(50));
  console.log(`Parse time: ${parseTime.toFixed(4)}s`);
  console.log(`Games per second: ${(parsed / parseTime).toFixed(2)}`);
  console.log(`Moves per second: ${(totalMoves / parseTime).toFixed(2)}`);
}

main().catch(console.error);

File: python/pgn_benchmark.py
#!/usr/bin/env python3
"""Pure PGN Parsing Benchmark - Python Implementation"""

import requests
import chess
import chess.pgn
import io
import time
import argparse
from concurrent.futures import ThreadPoolExecutor, as_completed

def fetch_archives(username: str):
    url = f"https://api.chess.com/pub/player/{username}/games/archives"
    headers = {"User-Agent": "ChessBenchmark/1.0"}
    resp = requests.get(url, headers=headers, timeout=30)
    resp.raise_for_status()
    return resp.json().get("archives", [])

def fetch_games(archive_url: str):
    headers = {"User-Agent": "ChessBenchmark/1.0"}
    resp = requests.get(archive_url, headers=headers, timeout=60)
    resp.raise_for_status()
    return resp.json().get("games", [])

def parse_game(pgn_str: str):
    if not pgn_str:
        return 0, 0
    try:
        pgn = chess.pgn.read_game(io.StringIO(pgn_str))
        if not pgn:
            return 0, 0
        board = pgn.board()
        move_count = 0
        position_count = 1
        for move in pgn.mainline_moves():
            board.push(move)
            move_count += 1
            position_count += 1
            _ = board.fen()
        return move_count, position_count
    except:
        return 0, 0

def main():
    parser = argparse.ArgumentParser(description='Pure PGN Parsing Benchmark')
    parser.add_argument('username', nargs='?', default='hikaru')
    parser.add_argument('games', nargs='?', type=int, default=1000)
    parser.add_argument('--workers', type=int, default=4)
    args = parser.parse_args()

    print(f"Python PGN Parsing Benchmark")
    print(f"=" * 50)
    print(f"Library: python-chess")
    print(f"Username: {args.username}")
    print(f"Max games: {args.games}")
    print(f"Threads: {args.workers}")
    print()

    print("Fetching games...")
    fetch_start = time.perf_counter()
    archives = list(reversed(fetch_archives(args.username)))
    
    all_pgns = []
    for archive_url in archives:
        if len(all_pgns) >= args.games:
            break
        games = fetch_games(archive_url)
        for g in games:
            if g.get("pgn"):
                all_pgns.append(g["pgn"])
        print(f"  Fetched {len(games)} games from {archive_url.split('/')[-2]}/{archive_url.split('/')[-1]}")

    all_pgns = all_pgns[:args.games]
    fetch_time = time.perf_counter() - fetch_start
    print(f"Fetched {len(all_pgns)} games in {fetch_time:.2f}s")
    print()

    print("Parsing PGNs...")
    parse_start = time.perf_counter()
    
    total_moves = 0
    total_positions = 0
    games_parsed = 0

    with ThreadPoolExecutor(max_workers=args.workers) as executor:
        futures = {executor.submit(parse_game, pgn): i for i, pgn in enumerate(all_pgns)}
        for i, future in enumerate(as_completed(futures)):
            moves, positions = future.result()
            if moves > 0:
                total_moves += moves
                total_positions += positions
                games_parsed += 1
            if (i + 1) % 100 == 0 or i == len(all_pgns) - 1:
                elapsed = time.perf_counter() - parse_start
                print(f"  Parsed {i + 1}/{len(all_pgns)} games ({(i + 1) / elapsed:.2f} games/sec)")

    parse_time = time.perf_counter() - parse_start

    print()
    print(f"Results")
    print(f"=" * 50)
    print(f"Games parsed: {games_parsed}")
    print(f"Total moves: {total_moves}")
    print(f"Total positions: {total_positions}")
    print()
    print(f"Performance")
    print(f"=" * 50)
    print(f"Parse time: {parse_time:.4f}s")
    print(f"Games per second: {games_parsed / parse_time:.2f}")
    print(f"Moves per second: {total_moves / parse_time:.2f}")

if __name__ == "__main__":
    main()
